ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,623 @@
+      telemetryContext: TelemetryContext,
+  ): CompletionStage[SubmissionResult] =
+    submit(
+      Submission.Config(
+        maxRecordTime = maxRecordTime,
+        submissionId = submissionId,
+        config = config,
+      )
+    )
+
+  override def currentHealth(): HealthStatus = Healthy
+
+  override def allocateParty(
+      hint: Option[Ref.Party],
+      displayName: Option[String],
+      submissionId: Ref.SubmissionId,
+  )(implicit
+      loggingContext: LoggingContext,
+      telemetryContext: TelemetryContext,
+  ): CompletionStage[SubmissionResult] =
+    submit(
+      Submission.AllocateParty(
+        hint = hint,
+        displayName = displayName,
+        submissionId = submissionId,
+      )
+    )
+
+  override def uploadPackages(
+      submissionId: Ref.SubmissionId,
+      archives: List[Archive],
+      sourceDescription: Option[String],
+  )(implicit
+      loggingContext: LoggingContext,
+      telemetryContext: TelemetryContext,
+  ): CompletionStage[SubmissionResult] =
+    submit(
+      Submission.UploadPackages(
+        submissionId = submissionId,
+        archives = archives,
+        sourceDescription = sourceDescription,
+      )
+    )
+
+  override def prune(
+      pruneUpToInclusive: Offset,
+      submissionId: Ref.SubmissionId,
+      pruneAllDivulgedContracts: Boolean,
+  ): CompletionStage[PruningResult] =
+    CompletableFuture.completedFuture(
+      PruningResult.ParticipantPruned
+    )
+
+  override def ledgerInitialConditions(): Source[LedgerInitialConditions, NotUsed] =
+    Source.single(
+      LedgerInitialConditions(
+        ledgerId = ledgerId,
+        config = Configuration(
+          generation = 1L,
+          timeModel = LedgerTimeModel.reasonableDefault,
+          maxDeduplicationTime = java.time.Duration.ofSeconds(maxDedupSeconds.toLong),
+        ),
+        initialRecordTime = Timestamp.now(),
+      )
+    )
+
+  override def stateUpdates(
+      beginAfter: Option[Offset]
+  )(implicit loggingContext: LoggingContext): Source[(Offset, Update), NotUsed] = {
+    // TODO for PoC purposes:
+    //   This method may only be called once, either with `beginAfter` set or unset.
+    //   A second call will result in an error unless the server is restarted.
+    //   Bootstrapping the bridge from indexer persistence is supported.
+    synchronized {
+      if (stateUpdatesWasCalledAlready)
+        throw new IllegalStateException("not allowed to call this twice")
+      else stateUpdatesWasCalledAlready = true
+    }
+    logger.info("Indexer subscribed to state updates.")
+    beginAfter.foreach(offset =>
+      logger.warn(
+        s"Indexer subscribed from a specific offset $offset. This offset is not taking into consideration, and does not change the behavior of the ReadWriteServiceBridge. Only valid use case supported: service starting from an already ingested database, and indexer subscribes from exactly the ledger-end."
+      )
+    )
+
+    queueSource.merge(source)
```

Comment:
interesting approach with merging streams, having a pass through lane, and getting the offset assignment form a atomic mutable!
not sure about production: there we might want to have more explicit offset assignement